### DISCUSSION TITLE
after_build_file 不能 hook proto 文件

### DISCUSSION THREAD

**Discussion by ivanallen** [xmake-io/xmake#4036] [Q&A]
### Xmake 版本

xmake v2.8.1+20230728

### 操作系统版本和架构

Linux 10-16-43-32 5.14.0-70.13.1.el9_0.x86_64 #1 SMP PREEMPT Wed May 25 21:01:57 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### 描述问题

after_build_file 不能 hook proto 文件，直接跳过了。

### 期待的结果

希望能 hook proto 文件，这样便自己去生成 proto 对应的 python 文件

### 工程配置

```lua
set_project("test")
set_languages("c++20")

set_version("1.0.0", {build = "%Y%m%d"})

add_requires("protobuf-cpp", {system = false})

target("demo")
    set_kind("binary")
    add_files("src/main.cc")
    add_files("src/demo.proto")
    add_rules("protobuf.cpp")
    add_packages("protobuf-cpp")
    add_packages("grpc")

    before_build_file(function (target, sourcefile, opt)
        print("before ====>", sourcefile, opt)
    end)

    after_build_file(function (target, sourcefile, opt)
        print("after ---->", sourcefile, opt)  -- 只打印了 main.cc，没有 demo.proto
    end)
```

### 附加信息和错误日志

![image](https://github.com/xmake-io/xmake/assets/12481610/e2184d34-76b9-40cd-8ed1-b4ef6c4d6caa)

[a comment by an automated account is not shown]

**Comment by waruqi** [comment]
target 的 after_build_file 只拦截已知的 cpp files，`.proto` 是 `protobuf.cpp` 的 rule 里面拦截处理的，它不属于内置的源文件。。你要 hook 也得走 rule ，然后通过 set_extensions 针对性 hook 

https://github.com/xmake-io/xmake/blob/05b62d2a23b5ad1cc8cf5df357f8c8cd0229e8c2/xmake/rules/protobuf/xmake.lua#L24

[a comment by an automated account is not shown]
